### ISSUE TITLE
Support arbitrary logical sequence lengths in SparsePiperAttention

### ISSUE THREAD

**Issue by Boffee** [Boffee/piper-kernels#54]
## Problem

`SparsePiperAttention` currently requires equal-shaped BF16 `[B, S, H, 128]` Q/K/V with `S % 64 == 0` and treats every supplied row as valid. This prevents direct use with ordinary H3 sequence lengths unless the caller pads the model sequence, and caller-owned padding would require separate validity metadata.

Dense Piper Attention already accepts a logical ragged sequence and handles its final partial tile internally. Sparse Piper should expose the same public behavior.

## Desired public contract

- Continue accepting equal-shaped BF16 `[B, S, H, 128]` self-attention Q/K/V.
- Allow arbitrary logical `S`; remove the public K64-alignment requirement.
- Return exactly `[B, S, H, 128]`.
- The caller supplies an unpadded logical sequence.
- Any K64 padding required for routing, quantization, or SM120 execution is internal.
- Do not require a public `valid_sequence_length` argument.

## Sparse-prefix semantics

`sparse_key_blocks` continues to count complete routeable K64 blocks:

```text
1 <= sparse_key_blocks <= floor(S / 64)
sparse prefix = [0, sparse_key_blocks * 64)
dense suffix  = [sparse_key_blocks * 64, S)
```

Any partial media block is therefore part of the always-dense suffix. DSA candidates remain limited to complete sparse-prefix K64 blocks.

## Ragged DSA behavior

DSA must produce routes for `ceil(S / 64)` query blocks. For the final partial Q64 block, only valid query rows may contribute to the query extrema and route selection.

DSA continues to inspect:

- all valid Q rows as queries;
- only complete sparse-prefix K blocks as candidates.

Internal padding must not affect DSA summaries or selected routes.

## Ragged attention behavior

For every valid query row, attention combines the DSA-selected sparse-prefix K/V blocks and every valid dense-suffix K/V row in one softmax.

The final partial dense K/V tile should match dense Piper boundary semantics:

- nonexistent key logits behave as negative infinity;
- nonexistent key probabilities are zero;
- nonexistent values do not contribute to the output.

The final partial query tile must produce only the `S` logical output rows.

## Preparation and quantization

Internal padding must not affect:

- final-query-block DSA summaries;
- Q quantization or Q scales;
- K/V means;
- K quantization or K scales;
- V quantization or final-tile V scales;
- attention logits, probabilities, or valid outputs.

## Required paths

The same logical contract should hold for:

- the portable sparse reference;
- the optimized SM120 sparse kernel;
- the already-quantized internal path;
- the ConvRot-to-sparse-Piper compiled fusion.

The ConvRot fusion should still match and fire for ragged logical lengths without requiring padded model hidden states.

## Acceptance coverage

Exercise lengths around K64 boundaries, including `64`, `65`, `127`, `128`, `129`, `181`, `191`, `192`, and `193`.

Verify that:

- output shape equals the logical input shape;
- outputs are finite;
- portable and SM120 results satisfy the existing quality tolerance;
- invalid internal Q storage cannot alter final-block DSA routes;
- the final dense K/V tail matches explicit masked-reference semantics;
- a keep ratio of `1.0` attends to every valid sequence row;
- the compiled ConvRot fusion matches the materialized ragged path and is actually selected.

## Non-goals

- Per-example lengths within one batch;
- arbitrary attention masks;
- partial routeable sparse-prefix blocks;
- causal or cross-attention support;
- smart spatial tiling.

The implementation may use masked loads, internally padded quantized buffers, or another equivalent mechanism. The externally observable requirement is that tensor dimension `S` is always the logical sequence length, matching dense Piper Attention.
